src/beamlime/core/system.py

```diff
@@ -1,79 +1,140 @@
     BeamlimeApplicationInterface,
 )
 from ..communication.broker import CommunicationBroker
+from ..config.preset_options import Timeout, WaitInterval
+from ..config.tools import dict_to_yaml
+from ..logging import LOG_LEVELS
+
+
+class ExitCode(Enum):
+    SUCCESS = 0
+    INTERRUPTED = 1
+    FAIL = 2
+
+
+def instantiate_system(
+    *,
+    name: str,
+    constructor: str,
+    specs: Optional[dict],
+    logger: Logger,
+    **unused_specs,
+) -> BeamlimeSystem:
+    from ..config.tools import import_object
+
+    constructor_obj = import_object(constructor)
+    logger.info(
+        "Unused information in the configuration: %s", dict_to_yaml(unused_specs)
+    )
+    return constructor_obj(name=name, logger=logger, **specs)
+
+
+def start_system(
+    config: Union[dict, str, None],
+    logger: Logger = None,
+    log_level: LOG_LEVELS = None,
+    verbose: bool = True,
+) -> BeamlimeSystem:
+    from ..config.loader import safe_load_config
+    from ..logging import safe_get_logger
+
+    _runtime_config = safe_load_config(config)  # TODO: configuration validation.
+    _logger = safe_get_logger(logger, verbose)
+    if log_level:
+        _logger.setLevel(log_level)
+
+    system_instance = instantiate_system(logger=_logger, **_runtime_config)
+
+    _logger.info(
+        "System instance created with this configuration: \n%s",
+        dict_to_yaml(_runtime_config),
+    )
+    return system_instance
 
 
 class BeamlimeSystem(BeamlimeApplicationInterface):
     """Manage various types of applications"""
 
     def __init__(
         self,
-        config: Union[dict, str, None] = None,
+        name: str = "Beamlime Dashboard",
         logger: Logger = None,
         save_log: bool = False,
+        log_dir: str = None,
+        timeout: float = Timeout.maximum,
+        wait_interval: float = WaitInterval.minimum,
+        applications: Iterable[Dict] = None,
+        communications: Iterable[Dict] = None,
+        subscriptions: Iterable[Dict] = None,
+        **unused_specs,
     ) -> None:
-        from copy import copy
-
-        from ..config.loader import safe_load_config
-        from ..config.tools import list_to_dict
-
-        self.user_config = safe_load_config(config)
-        self.config = copy(self.user_config)
-        self.system_config = self.config["general"]
-        self.app_configs = list_to_dict(self.config["data-stream"]["applications"])
-
         super().__init__(
-            app_name=self.system_config["system-name"],
-            broker=self.broker,
+            name=name,
             logger=logger,
-            timeout=self.system_config.get("timeout", 600),
-            wait_interval=self.system_config.get("wait-interval", 1),
+            timeout=timeout,
+            wait_interval=wait_interval,
         )
+
+        self._init_logger(save_log, log_dir)
+        self.broker = self._build_broker(communications, subscriptions)
+        self.applications = self._build_application(applications)
```

Comment:
It feels like there may be an advantage in invert the logic here. Right now, we pass things to `BeamlimeSystem`, which then constructs some sub-components. Dependency injection might be a better design choice here, making the entire thing less coupled, and more flexible (also for testing). I don't know if you have already considered that option?

Reply:
Yes, that was ultimately what I wanted to achieve, so I wanted to have an helper to build a pipeline.
And then it turned out to be more like a controller...
I will separate that part and inject them instead, and I thought I should remove that optional logger logic from the constructor as well so it makes sense.

Reply:
👍 (but maybe do that in a different PR?)